File: app/src/main/java/org/tasks/data/UserActivityDao.java
package org.tasks.data;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Update;
import com.todoroo.andlib.utility.DateUtilities;
import com.todoroo.astrid.data.Task;
import com.todoroo.astrid.helper.UUIDHelper;
import java.util.List;

@Dao
public abstract class UserActivityDao {

  @Insert
  public abstract void insert(UserActivity userActivity);

  @Update
  public abstract void update(UserActivity userActivity);

  @Delete
  public abstract void delete(UserActivity userActivity);

  @Query("SELECT * FROM userActivity WHERE target_id = :taskUuid ORDER BY created_at DESC ")
  public abstract List<UserActivity> getCommentsForTask(String taskUuid);

  @Query(
      "SELECT userActivity.* FROM userActivity INNER JOIN tasks ON tasks._id = :task WHERE target_id = tasks.remoteId")
  public abstract List<UserActivity> getComments(long task);

  @Query("SELECT * FROM userActivity")
  public abstract List<UserActivity> getComments();

  public void createNew(UserActivity item) {
    if (item.getCreated() == null || item.getCreated() == 0L) {
      item.setCreated(DateUtilities.now());
    }
    if (Task.isUuidEmpty(item.getRemoteId())) {
      item.setRemoteId(UUIDHelper.newUUID());
    }
    insert(item);
  }
}

File: app/src/main/java/org/tasks/data/UserActivityDao.kt
package org.tasks.data

import androidx.room.*
import com.todoroo.andlib.utility.DateUtilities
import com.todoroo.astrid.data.Task
import com.todoroo.astrid.helper.UUIDHelper
import kotlinx.coroutines.flow.Flow

@Dao
abstract class UserActivityDao {
    @Insert
    abstract suspend fun insert(userActivity: UserActivity)

    @Update
    abstract suspend fun update(userActivity: UserActivity)

    @Delete
    abstract suspend fun delete(userActivity: UserActivity)

    @Query("SELECT * FROM userActivity WHERE target_id = :taskUuid ORDER BY created_at DESC ")
    abstract fun watchComments(taskUuid: String): Flow<List<UserActivity>>

    @Query("SELECT userActivity.* FROM userActivity INNER JOIN tasks ON tasks._id = :task WHERE target_id = tasks.remoteId")
    abstract suspend fun getComments(task: Long): List<UserActivity>

    @Query("SELECT * FROM userActivity")
    abstract suspend fun getComments(): List<UserActivity>

    suspend fun createNew(item: UserActivity) {
        if (item.created == null || item.created == 0L) {
            item.created = DateUtilities.now()
        }
        if (Task.isUuidEmpty(item.remoteId)) {
            item.remoteId = UUIDHelper.newUUID()
        }
        insert(item)
    }
}